packages/mongodb-ts-autocomplete/src/index.ts

```diff
@@ -0,0 +1,279 @@
+        : `{}`;
+      return `      '${collectionName}': ShellAPI.Collection<${def}>;`;
+    });
+
+    return `{
+  ${collectionProperties.join('\n')}
+}`;
+  }
+}
+
+class ConnectionSchema {
+  private readonly databaseNames: string[];
+  private databaseName: string;
+  private readonly databaseSchemas: Record<string, DatabaseSchema>;
+
+  constructor() {
+    this.databaseNames = [];
+    this.databaseName = 'test';
+    this.databaseSchemas = Object.create(null);
+
+    this.addDatabase(this.databaseName);
+  }
+
+  setDatabaseName(databaseName: string) {
+    this.databaseName = databaseName;
+  }
+
+  getCurrentDatabaseName(): string {
+    return this.databaseName;
+  }
+
+  addDatabase(databaseName: string) {
+    this.databaseNames.push(databaseName);
+    this.databaseSchemas[databaseName] = new DatabaseSchema();
+  }
+
+  setDatabaseCollectionNames(databaseName: string, collectionNames: string[]) {
+    if (!this.databaseSchemas[databaseName]) {
+      throw new Error(`expected ${databaseName} to be known`);
+    }
+    this.databaseSchemas[databaseName].setCollectionNames(collectionNames);
+  }
+
+  addCollectionSchema(
+    databaseName: string,
+    collectionName: string,
+    collectionSchema: JSONSchema
+  ) {
+    if (!this.databaseSchemas[databaseName]) {
+      throw new Error(`expected ${databaseName} to be known`);
+    }
+    this.databaseSchemas[databaseName].setCollectionSchema(
+      collectionName,
+      collectionSchema
+    );
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const databaseProperties = this.databaseNames.map((databaseName) => {
+      const def = this.databaseSchemas[databaseName]
+        ? this.databaseSchemas[databaseName].toTypescriptTypeDefinition()
+        : `{}`;
+      return `      '${databaseName}': ShellAPI.Database & ${def}`;
+    });
+
+    return `{
+  ${databaseProperties.join('\n')}
+}`;
+  }
+}
+
+export default class MongoDBAutocompleter {
+  private readonly context: AutocompletionContext;
+  private connectionSchemas: Record<string, ConnectionSchema>;
+  private currentConnectionKey: string | undefined;
+  private readonly autocompleter: Autocompleter;
+
+  constructor({ context, autocompleterOptions }: MongoDBAutocompleterOptions) {
+    this.context = context;
+    this.autocompleter = new Autocompleter(autocompleterOptions);
+
+    this.connectionSchemas = Object.create(null);
+    this.connectionSchemas;
+
+    // TODO: set the @mongodb-js/mql-typescript definitions (will this be separate or part of shell-api?)
+    // these aren't globals because they are types, so probably only make sense
+    // in combination with either the shell-api or something like a document or
+    // aggregation editor
+
+    this.autocompleter.updateCode({
+      'bson-expressions.d.ts': loadBSONExpressions(),
+    });
+    this.autocompleter.updateCode({ 'shell-api.d.ts': loadShellAPI() });
+  }
+
+  setConnectionKey(connectionKey: string) {
```

Comment:
I have this idea of setting a connectionKey (which should probably be renamed to connectionId) and databaseName separate from the autocomplete() function, but I think a better API would be to have the autocomplete method just take the connectionId and databaseName as parameters. Then we don't need the concept of which connection and/or database is the current one.

Every time we autocomplete we already re-build current-globals.d.ts anyway because the current collection schema could have changed, so it might as well take that into account right there. We already have a transparent cache for the schemas regardless and the server schema is calculated and updated as one chunk.

Reply:
Ok this was dumb - cleaned up now.